WindowManager/app/src/main/java/com/example/windowmanagersample/DisplayFeaturesActivity.kt

```diff
@@ -0,0 +1,178 @@
+    private val displayFeatureViews = ArrayList<View>()
+    private val deviceStateChangeCallback = DeviceStateChangeCallback()
+    private val layoutStateChangeCallback = LayoutStateChangeCallback()
+
+    private lateinit var binding: ActivityDisplayFeaturesBinding
+
+    private var windowBackend: MidScreenFoldBackend? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityDisplayFeaturesBinding.inflate(layoutInflater)
+        val view = binding.root
+        setContentView(view)
+
+        windowBackend = getTestBackend()
+        windowManager = WindowManager(this, windowBackend)
+
+        if (windowBackend != null) {
+            binding.deviceStateToggleButton.visibility = View.VISIBLE
+            binding.deviceStateToggleButton.setOnClickListener {
+                windowBackend?.toggleDeviceHalfOpenedState()
+            }
+        }
+
+        stateLog.clear()
+        stateLog.append(getString(R.string.stateUpdateLog)).append("\n")
+
+        windowManager.registerDeviceStateChangeCallback(
+            mainThreadExecutor,
+            deviceStateChangeCallback
+        )
+
+        window.decorView.doOnLayout {
+            updateStateAndFeatureViews()
+        }
+    }
+
+    override fun onAttachedToWindow() {
+        super.onAttachedToWindow()
+        windowManager.registerLayoutChangeCallback(mainThreadExecutor, layoutStateChangeCallback)
+    }
+
+    override fun onDetachedFromWindow() {
+        super.onDetachedFromWindow()
+        windowManager.unregisterLayoutChangeCallback(layoutStateChangeCallback)
+    }
+
+    /**
+     * Update the device state and display feature positions. Needs to be called after window
+     * layout, so that view position in the window could be evaluated correctly.
+     */
+    internal fun updateStateAndFeatureViews() {
+        // Cleanup previously added feature views
+        for (featureView in displayFeatureViews) {
+            binding.featureContainerLayout.removeView(featureView)
+        }
+        displayFeatureViews.clear()
+
+        // Update the UI with the current state
+        val stateStringBuilder = StringBuilder()
+        // Update the current state string
+        stateStringBuilder.append(getString(R.string.deviceState))
+            .append(": ")
+            .append(windowManager.deviceState)
+            .append("\n")
+
+        stateStringBuilder.append(getString(R.string.windowLayout))
+            .append(": ")
+            .append(windowManager.windowLayoutInfo)
+
+        // Add views that represent display features
+        for (displayFeature in windowManager.windowLayoutInfo.displayFeatures) {
+            val lp = getLayoutParamsForFeatureInFrameLayout(
+                displayFeature,
+                binding.featureContainerLayout
+            ) ?: continue
+
+            // Make sure that zero-wide and zero-high features are still shown
+            if (lp.width == 0) {
```

Comment:
You should probably check the actual view's width, rather than the lp value.

Reply:
LPs should only be read and intepreted by the parent, since they can mean different things to different ViewGroups. For instance: a value of `0` width in `ConstraintLayout` means 'match_constraint', not 0px. The final laid out width is the only source of truth for any consumers.

Use `doOnLayout()`?